docs/proposals/20191017-control-plane.md

```diff
@@ -0,0 +1,544 @@
+
+    // CreateClusterError indicates that an error was encountered
+    // when trying to create the cluster.
+    CreateClusterError ClusterStatusError = "CreateError"
+
+    // UpdateClusterError indicates that an error was encountered
+    // when trying to update the cluster.
+    UpdateClusterError ClusterStatusError = "UpdateError"
+
+    // DeleteClusterError indicates that an error was encountered
+    // when trying to delete the cluster.
+    DeleteClusterError ClusterStatusError = "DeleteError"
+)
+```
+
+With the following validations:
+
+- If `ControlPlane.Spec.KubeadmConfigSpec` does not define external etcd (webhook):
+  - `ControlPlane.Spec.Replicas` in [1, 3, 5, 7]
+- `ControlPlane.Spec.Version != ""` (openapi)
+
+And the following defaulting:
+
+- `ControlPlaneSpec.Replicas: 1` (openapi or webhook)
+
+#### Modifications required to existing API Types
+
+- Add `Cluster.Spec.ControlPlaneRef` defined as:
+
+```go
+    // ControlPlaneRef is an optional reference to a provider-specific resource that holds
+    // the details for provisioning the Control Plane for a Cluster
+    // +optional
+    ControlPlaneRef *corev1.ObjectReference `json:"controlPlaneRef,omitempty"`
+```
+
+- Add `Cluster.Status.ControlPlaneReady` defined as:
+
+```go
+    // ControlPlaneReady defines if the control plane is ready
+    // +optional
+    ControlPlaneReady bool `json:"controlPlaneReady"`
+```
+
+#### Behavioral Changes from v1alpha2
+
+- If Cluster.Spec.ControlPlaneRef is set:
+  - [Status.ControlPlaneInitialized](https://github.com/kubernetes-sigs/cluster-api/issues/1243) is set based on the value of Status.Initialized for the referenced resource.
+  - Status.ControlPlaneReady is set based on the value of Status.Ready for the referenced resource, this field is intended to eventually replace Status.ControlPlaneInitialized as a field that will be kept up to date instead of set only once.
+- Current behavior will be preserved if `Cluster.Spec.ControlPlaneRef` is not set.
+
+#### Behaviors
+
+##### Create
+
+- After a ControlPlane object is created, it must bootstrap a control plane with a given number of replicas.
+- If an error occurs, `ControlPlane.Status.ErrorStatus` and `ControlPlane.Status.ErrorMessage` are populated.
+- Cannot create an even number of stacked control planes or anything other than 1, 3, 5, or 7
+- Can create an arbitrary number of control planes if etcd is external to the control plane
+- Creating a ControlPlane with > 1 replicas is equivalent to creating a ControlPlane with 1 replica followed by scaling the ControlPlane to 3 replicas
+- The kubeadm bootstrapping configuration provided via `ControlPlane.Spec.KubeadmConfigSpec` should specify the `InitConfiguration`, `ClusterConfiguration`, and `JoinConfiguration` stanzas, and the ControlPlane controller will be responsible for splitting the config and passing it to the underlying Machines created as appropriate.
```

Comment:
Is this any different than how this is handled today? If not, why call it out explicitly?

Reply:
Today each Machine has a reference to it's own KubeadmConfig, in this case we are taking a singular KubeadmConfig and using it to create the individual KubeadmConfigs for each Machine similar to how the MachineDeployment uses a KubeadmConfigTemplate.